File: app/src/main/java/com/mgaetan89/showsrage/fragment/ScheduleFragment.java
package com.mgaetan89.showsrage.fragment;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.mgaetan89.showsrage.R;
import com.mgaetan89.showsrage.activity.MainActivity;
import com.mgaetan89.showsrage.adapter.SchedulePagerAdapter;
import com.mgaetan89.showsrage.model.Schedule;
import com.mgaetan89.showsrage.model.Schedules;
import com.mgaetan89.showsrage.network.SickRageApi;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import retrofit.Callback;
import retrofit.RetrofitError;
import retrofit.client.Response;

public class ScheduleFragment extends Fragment implements Callback<Schedules> {
	@Nullable
	private SchedulePagerAdapter adapter = null;

	@NonNull
	private final List<ArrayList<Schedule>> schedule = new ArrayList<>();

	@NonNull
	private final List<String> sections = new ArrayList<>();

	@Nullable
	private TabLayout tabLayout = null;

	@Nullable
	private ViewPager viewPager = null;

	public ScheduleFragment() {
	}

	@Override
	public void failure(RetrofitError error) {
		error.printStackTrace();
	}

	@Override
	public void onActivityCreated(@Nullable Bundle savedInstanceState) {
		super.onActivityCreated(savedInstanceState);

		FragmentActivity activity = this.getActivity();

		if (activity instanceof MainActivity) {
			((MainActivity) activity).displayHomeAsUp(false);
			activity.setTitle(R.string.schedule);
		}

		SickRageApi.Companion.getInstance().getServices().getSchedule(this);

		this.tabLayout = (TabLayout) this.getActivity().findViewById(R.id.tabs);

		if (this.viewPager != null && this.tabLayout != null) {
			this.tabLayout.setupWithViewPager(this.viewPager);
		}
	}

	@Nullable
	@Override
	public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
		View view = inflater.inflate(R.layout.fragment_schedule, container, false);

		if (view != null) {
			this.viewPager = (ViewPager) view.findViewById(R.id.schedule_pager);

			if (this.viewPager != null) {
				this.adapter = new SchedulePagerAdapter(this.getChildFragmentManager(), this.sections, this.schedule);

				this.viewPager.setAdapter(this.adapter);
			}
		}

		return view;
	}

	@Override
	public void onDestroy() {
		this.schedule.clear();
		this.sections.clear();

		super.onDestroy();
	}

	@Override
	public void onDestroyView() {
		this.tabLayout = null;
		this.viewPager = null;

		super.onDestroyView();
	}

	@Override
	public void success(Schedules schedules, Response response) {
		if (schedules != null) {
			Map<String, ArrayList<Schedule>> data = schedules.getData();

			if (data != null) {
				String statuses[] = {"missed", "today", "soon", "later"};

				for (String status : statuses) {
					if (data.containsKey(status)) {
						ArrayList<Schedule> scheduleForStatus = data.get(status);

						if (!scheduleForStatus.isEmpty()) {
							this.schedule.add(scheduleForStatus);

							if (this.isAdded()) {
								this.sections.add(this.getString(getSectionName(status)));
							} else {
								this.sections.add(status);
							}
						}
					}
				}

				if (this.adapter != null) {
					this.adapter.notifyDataSetChanged();

					if (this.tabLayout != null) {
						this.tabLayout.setTabsFromPagerAdapter(this.adapter);
						this.tabLayout.setVisibility(View.VISIBLE);
					}
				}
			}
		}
	}

	@StringRes
	/* package */ static int getSectionName(String sectionId) {
		if (sectionId != null) {
			String normalizedSectionId = sectionId.toLowerCase();

			switch (normalizedSectionId) {
				case "later":
					return R.string.later;

				case "missed":
					return R.string.missed;

				case "soon":
					return R.string.soon;

				case "today":
					return R.string.today;
			}
		}

		return 0;
	}
}

File: app/src/main/kotlin/com/mgaetan89/showsrage/fragment/ScheduleFragment.kt
package com.mgaetan89.showsrage.fragment

import android.os.Bundle
import com.mgaetan89.showsrage.R
import com.mgaetan89.showsrage.activity.MainActivity
import com.mgaetan89.showsrage.adapter.SchedulePagerAdapter
import com.mgaetan89.showsrage.extension.clearSchedule
import com.mgaetan89.showsrage.extension.getScheduleSections
import com.mgaetan89.showsrage.extension.saveSchedules
import com.mgaetan89.showsrage.model.Schedules
import com.mgaetan89.showsrage.network.SickRageApi
import io.realm.Realm
import kotlinx.android.synthetic.main.fragment_tabbed.swipe_refresh
import retrofit.Callback
import retrofit.RetrofitError
import retrofit.client.Response

class ScheduleFragment : TabbedFragment(), Callback<Schedules> {
	private val sectionIds = mutableListOf<String>()
	private val sectionLabels = mutableListOf<String>()

	override fun failure(error: RetrofitError?) {
		this.swipe_refresh?.isRefreshing = false

		error?.printStackTrace()
	}

	override fun onActivityCreated(savedInstanceState: Bundle?) {
		super.onActivityCreated(savedInstanceState)

		val activity = this.activity

		if (activity is MainActivity) {
			activity.displayHomeAsUp(false)
			activity.setTitle(R.string.schedule)
		}

		val sections = Realm.getDefaultInstance().use(Realm::getScheduleSections)

		this.setSections(sections)
		this.onRefresh()
	}

	override fun onDestroy() {
		this.sectionIds.clear()
		this.sectionLabels.clear()

		super.onDestroy()
	}

	override fun onRefresh() {
		this.swipe_refresh?.isRefreshing = true

		SickRageApi.instance.services?.getSchedule(this)
	}

	override fun success(schedules: Schedules?, response: Response?) {
		this.swipe_refresh?.isRefreshing = false

		val data = schedules?.data ?: return

		this.sectionIds.clear()
		this.sectionLabels.clear()

		this.setSections(data.keys.filter {
			data[it]?.isNotEmpty() ?: false
		})

		val realm = Realm.getDefaultInstance()
		realm.clearSchedule()

		data.forEach {
			if (it.value.isNotEmpty()) {
				realm.saveSchedules(it.key, it.value)
			}
		}

		realm.close()
	}

	override fun getAdapter() = SchedulePagerAdapter(this.childFragmentManager, this.sectionIds, this.sectionLabels)

	override fun useSwipeToRefresh() = true

	private fun setSections(sections: List<String>) {
		val statuses = listOf("missed", "today", "soon", "later")

		statuses.forEach {
			if (sections.contains(it)) {
				this.sectionIds.add(it)
				this.sectionLabels.add(if (this.isAdded) {
					this.getString(getSectionName(it))
				} else {
					it
				})
			}
		}

		this.updateState(this.sectionIds.isEmpty())
	}

	companion object {
		fun getSectionName(sectionId: String?) = when (sectionId?.toLowerCase()) {
			"later" -> R.string.later
			"missed" -> R.string.missed
			"soon" -> R.string.soon
			"today" -> R.string.today
			else -> 0
		}

		fun newInstance() = ScheduleFragment()
	}
}
